### DISCUSSION TITLE
Roadmap for upgrading to AWS provider >=4.x

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#299] [Help] [answered]
Where can I find information about your plan for a future upgrade to AWS provider >= 4.x?

**Comment by pete0emerson** [comment]
From a Gruntwork customer:

I too would like to know but I know it's not an easy answer to give.
I don't wanna speak for gruntworks about their plan or thoughts...
However, as someone who has worked a lot with terraform without terragrunt before, the AWS Provider >= 4.x is a huge foundational type of shift because on how hashicorp is trying to push S3 resource provisioning. Terraform is trying to consolidate all the issues on their github aws provider repo here in the [4.x upgrade](https://github.com/hashicorp/terraform-provider-aws/issues/23106#issuecomment-1035548640) on s3 bucket resources but many people have voiced their displeasure with the upgrade. 4.0.0 was released on Feb 10th 2022, which took many people by surprise.
Since all the states, logs, and configuration are being saved in those S3 resources... they can't screw up an upgrade for it and terraform's provider [upgrade process is sorta overwhelming](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade). It's really about that state management and it's literally the foundation of all terraform... that has to be a 100% clean update.
If the 4.x upgrade wasn't centered around s3 bucket resource, I think they would have had something out by now for an upgrade path, but since it is s3 bucket resources... they gotta get it right.
Yori has put in a lot of stopgaps in the current terragrunt private modules to keep things running so far. I anticipate that they are working on trying to get an upgrade path figured out but it's not gonna be easy.

**Comment by pete0emerson** [comment, marked as answer]
Solid thoughts above, thank you.

Our current solution is to lock the provider to < 4.x, but that can't stay that way forever!

Currently we don't publish a roadmap of when features and improvements like this will be done (an unsatisfactory answer, I know). We have talked about that, internally, but we're not there, yet.

**Reply by yorinasub17** [reply]
I would also like to add that we are actively working on the transition. Note that part of the reason we are a bit slower with the transition is exactly as mentioned above, where we are very mindful of the migration path for customers.

We currently have [an active PR](https://github.com/gruntwork-io/terraform-aws-security/pull/612) for the S3 bucket transition for one of our core modules (`private-s3-bucket`) which is used everywhere we need S3 buckets, and one of the key blockers for that PR is identifying the migration steps for a customer upgrading. You can subscribe to that PR to follow along with the conversation, and also be notified when we ultimately release this first step.

**Reply by LuigiClemente-Awin** [reply]
I see the new version of the security module is marked as compatible with AWS provider 4.6.0, but I see no update in this thread. 
Are you still working on other modules before officially releasing this?

**Reply by rhoboat** [reply]
Thanks for your interest! I know this is taking us a while to release due to a load of other priorities. The provider 4.x update is in active development, but we can’t make any commitments on when this will be done. We plan on providing an announcement in the newsletter and/or blog post when we complete the work.

**Comment by fsanzdev** [comment]
AWS provider versions =>4.9.0 retain the same configurations parameters of the `aws_s3_bucket` (they basically undid the changes) so I guess it's secure to use those versions https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade#changes-to-s3-bucket-drift-detection

Could you please confirm it?

**Reply by rhoboat** [reply]
Thanks for your interest! I know this is taking us a while to release due to a load of other priorities. The provider 4.x update is in active development, and the provider versions keep updating and rolling various things back and forth. Bear with us as we work through it all and get our modules updated. We can’t make any commitments on when this will be done. We plan on providing an announcement in the newsletter and/or blog post when we complete the work.

**Reply by fsanzdev** [reply]
Thanks for the quick reply, but the comment was related to the fact that with =>4.9.0 provider, you won't need to update your code (they have moved the breaking changes to 5.x) so it should be safe to use =>4.9.0 with the actual code according to the Hashicorp link above

**Reply by LuigiClemente-Awin** [reply]
Yes, as @fsanzdev said, please try removing version locks <4.0.x and replacing them with  =>4.9.0.

**Reply by LuigiClemente-Awin** [reply]
For example, I still see locks with version = "< 4.0" in terraform-aws-lambda/blob/master/modules/lambda/main.tf. This is blocking the upgrade for me.

**Reply by fsanzdev** [reply]
@rhoboat are you considering this option?

**Comment by LuigiClemente-Awin** [comment]
An update on this would be much appreciated, please.
